src/provider/index.ts

```diff
@@ -1,132 +1,114 @@
-
-    /**
-     * The reason that an app wasn't selected.
-     *
-     * Only specified when success is false.
-     */
-    reason?: string;
-
-    /**
-     * Determines the future behaviour of this intent
-     */
-    defaultAction: DefaultAction;
+@injectable()
+export class Main {
+    private _config = null;
+
+    @inject(Inject.APP_DIRECTORY)   private _directory!: AppDirectory;
+    @inject(Inject.MODEL)           private _model!: Model;
+
+    @inject(Inject.CONTEXT_HANDLER) private _contexts!: ContextHandler;
+    @inject(Inject.INTENT_HANDLER)  private _intents!: IntentHandler;
+
+    private apiHandler: APIHandler<APITopic>;
+
+    constructor() {
+        this.apiHandler = new APIHandler();
+    }
+
+    public async register(): Promise<void> {
+        Object.assign(window, {
+            main: this,
+            config: this._config,
+            directory: this._directory,
+            model: this._model,
+            contexts: this._contexts,
+            intents: this._intents
+        });
+
+        // Wait for creation of any injected components that require async initialization
+        await Injector.initialized;
+
+        // Current API
+        this.apiHandler.registerListeners<API>({
+            [APITopic.OPEN]: this.onOpen.bind(this),
+            [APITopic.FIND_INTENT]: this.findIntent.bind(this),
+            [APITopic.FIND_INTENTS_BY_CONTEXT]: this.findIntentsByContext.bind(this),
+            [APITopic.BROADCAST]: this.broadcast.bind(this),
+            [APITopic.RAISE_INTENT]: this.raiseIntent.bind(this)
+        });
+
+        console.log('Service Initialised');
+    }
+
+    private async onOpen(payload: OpenPayload): Promise<void> {
+        const appInfo: Application|null = await this._directory.getAppByName(payload.name);
+
+        if (appInfo) {
+            const app = await this._model.findOrCreate(appInfo, {prefer: FindFilter.WITH_CONTEXT_LISTENER});
+
+            if (payload.context) {
+                await this._contexts.send(app, payload.context);
+            }
+        } else {
+            throw new Error(`No app in directory with name: ${payload.name}`);
+        }
+    }
+
+    private async findIntent(payload: FindIntentPayload): Promise<AppIntent> {
+        let apps;
+        if (payload.intent) {
+            apps = await this._directory.getAppsByIntent(payload.intent);
+        } else {
+            apps = await this._directory.getAllApps();
+        }
+
+        return {
+            intent: {
+                name: payload.intent,
+                displayName: payload.intent
+            },
+            apps
+        };
+    }
+
+    private async findIntentsByContext (payload: FindIntentsByContextPayload): Promise<AppIntent[]> {
+        return [];
+    }
+
+    private async broadcast(payload: BroadcastPayload): Promise<void> {
+        await this._contexts.broadcast(payload.context);
+    }
+
+    private async raiseIntent(payload: RaiseIntentPayload): Promise<IntentResolution> {
```

Comment:
Shouldn't this also handle payload.target in some way?

Reply:
Yeah. This is still just a skeleton at this point - functionality is less than in current develop.

In the interests of unblocking things, will split this up into separate stories rather than trying to re-build all the current feature set into this structure within a single story. If there's no objections to that plan, of course - but makes sense to me.

Reply:
I'd say there's no harm in propagating the optional param down to the point where it would be used, even if it just gets ignored with a comment. Targets, while optional, are a pretty significant part of the api spec and I think the skeleton should at the very least acknowledge their existence.